src/basic/utf8.c

```diff
@@ -389,11 +389,17 @@ char *utf16_to_utf8(const char16_t *s, size_t length /* bytes! */) {
         const uint8_t *f;
         char *r, *t;
 
+        if (length == 0)
+                return new0(char, 1);
+
         assert(s);
 
+        if (length == SIZE_MAX)
+                length = char16_strlen(s) * 2;
+
         /* Input length is in bytes, i.e. the shortest possible character takes 2 bytes. Each unicode character may
          * take up to 4 bytes in UTF-8. Let's also account for a trailing NUL byte. */
-        if (length * 2 < length)
+        if (length > (SIZE_MAX - 1) / 2)
```

Comment:
Shouldn't you do the overflow check before multiplying the length?

Reply:
this is the overflow check for the multiplication below. But yeah, we should add one for the multiplication above too, to be fully correct...